documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,441 @@
+* *Pipe*: The duplex communication construct this protocol is communicated over.  This is a Unix Domain Socket on *nix systems and a Named Pipe on Windows.
+* *IPC Message*: The base unit of communication over the Diagnostic IPC Protocol. Is made up of a Header and a Payload.
+* *Header*: A struct containing version and metadata.
+* *Payload*: An opaque chunk of data that is Command specific.
+* *Optional Continuation*: The reuse of the pipe for application specific communication. This communication does not need to adhere to any requirements listed in this spec, e.g., this could be a stream of custom encoded data that is Command specific.
+
+### General Flow
+
+All communication with the Diagnostic Server will begin with a Diagnostic IPC Message sent from the client to the server.  The server will respond with a Diagnostic IPC Message.  After this, the client and runtime _may_ reuse the Pipe for any Command specific communication which is referred to as an Optional Continuation.
+
+![Generic Flow](ipc-protocol-genericflow.svg)
+
+```
+runtime <- client : [ Header ][ Payload ]
+runtime -> client : [ Header ][ Payload ]
+optional:
+    runtime <-> client : [ Optional Continuation ]
+connection closed
+```
+
+Example flow for EventPipe:
+```
+runtime <- client : [ ver; size; EventPipe CollectTracing ][ stream config struct  ] <- Diagnostic IPC Message
+runtime -> client : [ ver; size; EventPipe TracingStarted ][ stream started struct ] <- Diagnostic IPC Message
+runtime -> client : [ stream of netperf data ]                                       <- Optional Continuation
+
+// stop message is sent on another connection
+
+connection closed
+```
+
+Example flow for hypothetical application with custom communication protocol:
+```
+runtime <- client : [ ver; size; start_app_session_command    ][ app config struct      ] <- Diagnostic IPC Message
+runtime -> client : [ ver; size; app_session_started_response ][ session started struct ] <- Diagnostic IPC Message
+
+Optional Continuation using app specific protocol
+
+runtime <- client : [ command specific protocol ]                                         <- Optional Continuation
+runtime -> client : [ command specific protocol ]                                         <- Optional Continuation
+Eventually...
+runtime <- client : [ command specific protocol (close connection) ]                      <- Optional Continuation
+connection closed
+```
+
+### Transport
+
+The protocol will be communicated over a platform-specific transport.  On Unix/Linux based platforms, a Unix Domain Socket will be used, and on Windows, a Named Pipe will be used.
+
+#### Naming and Location Conventions
+
+Unix Domain Sockets:
+> The socket is created in the `tmp` dir.  This will be determined by the output of the `Path.GetTempPath()` function in managed code.  On Mac, this is typically an application group specific temp directory which can be found in the `$TMPDIR` environment variable.
+
```

Comment:
@jorive can you shed some light on the design decisions that led to the current implementation (which is what is reflected in the spec currently).  The code changes to implement this spec might be the perfect opportunity to make any changes to this that we want.  I will have to research a little to answer all your questions, Noah.

Reply:
@mikem8361 probably also has some historical info because the current scheme was duplicated from the debugger.

Reply:
I see you updated the spec to use the timestamp everywhere now, have you confirmed we didn't break the mac ApplicationGroup scenario or that is still a TODO?

Reply:
I double checked the logic and found that the disambiguation key is _always_ a timestamp.  The AppGroupID is only used to determine the temp dir, and I don't believe that logic is being used.  On my Macbook Pro, the pipes are always created in the `$TMPDIR` location.

I haven't had a chance to dig too much deeper into what's going on here.

disambiguation key -> https://github.com/dotnet/coreclr/blob/master/src/pal/src/thread/process.cpp#L2166

getting file location -> https://github.com/dotnet/coreclr/blob/master/src/pal/src/thread/process.cpp#L2298

Reply:
@mikem8361 - do you recall what scenario we should be testing here? The proposal is to stop writing sockets into the AppGroupContainer location and instead to always use $TMPDIR or /tmp. I'd rather not just wing it ; )

Reply:
@mikem8361 - I think you meant to respond to this comment? Copy/pasting from the other comment:

> On MacOS, it needs to be written in the application group container. See this code in the PAL for details: https://github.com/dotnet/coreclr/blob/c03f31fbc4105cbea4053409f91bc25040093d20/src/pal/src/thread/process.cpp#L2336
> I don't think we can always use $TMPDIR. I think we need to follow how the debugger transport pipe names are built. I don't know all the details; it was @odhanson that did all the application group work.


Thanks makes sense. Guess we are stuck with it.

Reply:
@josalem , I haven't read the design doc - just responding to this comment. 
The reason you only see the $TMPDIR being used, is because you are not trying to debug an application running inside a MacOS sandbox. The changes I made, were done in order to support debugging sandboxed apps. In this case, the $TMPDIR is different for each application and cannot be used as a common location for both the host and the target processes to share the IPC filename. The Application Group folder is a common folder that the app will have access to, and the host debugger (which runs outside of the sandbox) will also be able to access.

Having said that, I am fine with changing the form of the IPC, as long as we continue to use the APP group folder when we are running in a sandbox

Reply:
So long as that code path is _only_ hit when run inside a sandbox, I think it may be fine to leave the logic as is.  In that case, the tools will need to be updated to account for that case, as right now, they only look in `Path.GetTempPath()` which looks first in `$TMPDIR` then `/tmp`.  I'll modify the spec to reflect the sandbox code path as a rare variant, and create an issue to make the tools able to look in the sandbox dirs.

Reply:
Hey, we worked so hard to add support for debugging sandbox applications - Don't break us :)

Right now, the tools use dbgshim to obtain the path to the IPC file. So that enabled both sides to come up with the correct path. I understand that this proposal is planning to change that right ?

Reply:
Ah, by "tools" I'm referring to the new CLI tools in the dotnet/diagnostics repo which are currently only aware of pipes that exist inside `$TMPDIR` or `/tmp`.  I'm proposing that we don't change the pipe location logic in the PAL.  No sense breaking things other features rely on 😄!

Reply:
Oden, this new IPC protocol doesn't affect the debugger transport channel or all the work you did to support MacOS containers. This is a new channel allowing some of new tools like trace events, profiler attach and dump generation to communicate with the runtime. We just want to make sure it also runs in a MacOS container.

Reply:
> . I'll modify the spec to reflect the sandbox code path as a rare variant, and create an issue to make the tools able to look in the sandbox dirs.

The problem is that the tools will need to have some configuration to tell them which application group to use - this is something that can change from app to app.

Reply:
> Oden, this new IPC protocol doesn't affect the debugger transport channel or all the work you did to support MacOS containers. This is a new channel allowing some of new tools like trace events, profiler attach and dump generation to communicate with the runtime. We just want to make sure it also runs in a MacOS container.

Got it. Yeah, if we take the sandbox scenario into consideration while designing this - it would be great. I am sure as time passes, .net core will be used in more apps and sandboxing is a requirement for releasing to Apple store. Esp. with .Net 5 plans to allow xamarin with .net core runtime to work side by side.

Reply:
Tracking in #277 !